internal/sessioninterestmanager/sessioninterestmanager.go

```diff
@@ -40,26 +52,68 @@ func (sim *SessionInterestManager) RemoveSession(ses uint64) []cid.Cid {
 	sim.lk.Lock()
 	defer sim.lk.Unlock()
 
-	sim.wanted.RemoveSession(ses)
-	return sim.interested.RemoveSession(ses)
+	// The keys that no session is interested in
+	deletedKs := make([]cid.Cid, 0)
+
+	// For each known key
+	for c := range sim.wants {
+		// Remove the session from the list of sessions that want the key
+		delete(sim.wants[c], ses)
+
+		// If there are no more sessions that want the key
+		if len(sim.wants[c]) == 0 {
+			// Clean up the list memory
+			delete(sim.wants, c)
+			// Add the key to the list of keys that no session is interested in
+			deletedKs = append(deletedKs, c)
+		}
+	}
+
+	return deletedKs
 }
 
 // When the session receives blocks, it calls RemoveSessionWants().
-func (sim *SessionInterestManager) RemoveSessionWants(ses uint64, wants []cid.Cid) {
+func (sim *SessionInterestManager) RemoveSessionWants(ses uint64, ks []cid.Cid) {
 	sim.lk.Lock()
 	defer sim.lk.Unlock()
 
-	sim.wanted.RemoveSessionKeys(ses, wants)
+	// For each key
+	for _, c := range ks {
+		// If the session wanted the block
+		if wanted, ok := sim.wants[c][ses]; ok && wanted {
+			// Mark the block as unwanted
+			sim.wants[c][ses] = false
+		}
+	}
 }
 
 // When a request is cancelled, the session calls RemoveSessionInterested().
 // Returns the keys that no session is interested in any more.
-func (sim *SessionInterestManager) RemoveSessionInterested(ses uint64, wants []cid.Cid) []cid.Cid {
+func (sim *SessionInterestManager) RemoveSessionInterested(ses uint64, ks []cid.Cid) []cid.Cid {
```

Comment:
Something here is very strange. I think the meaning of "session interest" has changed over time. In the past, I believe it included everything the session _wanted_, but hadn't necessarily asked for.

I prefer the new interpretation "things the session is interested in hearing about", but I think we've made a mistake: we're canceling blocks when we become uninterested in them.

Shouldn't we be canceling blocks when they become _unwanted_?

Reply:
It's true it's a little convoluted. We actually want to do both:
- When the client calls Session.GetBlocks() the session registers **interest** in those blocks
- When we receive blocks / HAVEs / DONT_HAVEs we want to know which sessions are **interested** in them
- When the session receives a block it no longer **wants** the block, and sends a cancel.
  Note that the session is still *interested* in the block (but doesn't *want* the block), because the session wants to add peers who had the block to the session, so as to query them for other blocks that the session wants
- When the client cancels Session.GetBlocks() the session is no longer **interested** in those blocks, and should also send a cancel for them
- When the session shuts down, it is no longer **interested** in the remaining blocks and sends cancels for them

Reply:
My point is that this function returns CIDs (for cancellation) that _no_ sessions are interested in. However, we really want to return CIDs that no session _wants_.

I think we're technically fine because:

1. We cancel blocks when we receive them.
2. We'll lose interest in blocks when GetBlocks is canceled.

But it's still a bit strange.

Reply:
I see what you're saying, it's a little strange that we don't just cancel something that no one wants.

When we receive a block we send a cancel to the peer, and mark it as unwanted in the SessionInterestManager. We don't remove it entirely from the SessionInterestManager because if other peers subsequently respond with a HAVE or a block, we want to be able to tell the Session so that it can expand its pool of peers who may have blocks its interested in (this is particularly important in the discovery phase).

On the other hand if the request is cancelled, then the Session is not interested at all in hearing about subsequent blocks or HAVEs for the CIDs in the request, so in that case we remove the CIDs entirely from the SessionInterestManager.

Maybe the naming could be a clearer, I'm not sure that wanted / interested fully captures the concepts.